File: app/src/main/java/com/lence/penguinsvskillerwhales/life/Breeding.java
package com.lence.penguinsvskillerwhales.life;

import android.graphics.Point;

import com.lence.penguinsvskillerwhales.model.Organism;
import com.lence.penguinsvskillerwhales.utils.Constants;
import com.lence.penguinsvskillerwhales.view.KotlinField;
import com.lence.penguinsvskillerwhales.view.UpdatePresenter;

import java.util.ArrayList;
import java.util.Random;

public class Breeding {
    private final KotlinField mField;
    private final Point mCurrent;
    private UpdatePresenter mUpdatePresenter;
    private Organism mOrganism;

    Breeding(KotlinField mField, Point mCurrent, UpdatePresenter mUpdatePresenter, Organism mOrganism) {
        this.mField = mField;
        this.mCurrent = mCurrent;
        this.mUpdatePresenter = mUpdatePresenter;
        this.mOrganism = mOrganism;
    }

    public void execute() {
        ArrayList<Integer> checkList = new ArrayList<>();
        //free directions list
        for (int direction : Constants.directions) {
            if (findDirection(direction)) {
                checkList.add(direction);
            }
        }
        if (checkList.size() > 0) {
            int randomNumber;
            if (checkList.size() == 1) {
                randomNumber = 0;
            } else {
                Random random = new Random();
                randomNumber = random.nextInt(checkList.size() - 1);
            }
            //creation of a new organism
            mField.set(KotlinField.getPointByDirection(mCurrent.x, mCurrent.y,checkList.get(randomNumber)), mOrganism);
        }
        mUpdatePresenter.update(mField);
    }

    private boolean findDirection(int direction) {
        Point p = KotlinField.getPointByDirection(mCurrent.x, mCurrent.y, direction);

        return mField.isInField(p.x, p.y) && (mField.get(p.x, p.y) == null);

    }
}

File: app/src/main/java/com/lence/penguinsvskillerwhales/life/Breeding.kt
package com.lence.penguinsvskillerwhales.life

import android.graphics.Point
import com.lence.penguinsvskillerwhales.model.Organism
import com.lence.penguinsvskillerwhales.utils.Constants
import com.lence.penguinsvskillerwhales.view.Field
import com.lence.penguinsvskillerwhales.view.UpdatePresenter
import java.util.*

class Breeding internal constructor(private val mField: Field,
                                    private val mCurrent: Point,
                                    private val mUpdatePresenter: UpdatePresenter,
                                    private val mOrganism: Organism) {

    fun execute() {
        val checkList = ArrayList<Int>()
        //free directions list
        for (direction in Constants.directions) {
            if (findDirection(direction)) {
                checkList.add(direction)
            }
        }
        if (checkList.size > 0) {
            val randomNumber: Int
            if (checkList.size == 1) {
                randomNumber = 0
            } else {
                val random = Random()
                randomNumber = random.nextInt(checkList.size - 1)
            }
            //creation of a new organism
            mField[Field.getPointByDirection(mCurrent.x, mCurrent.y,
                    checkList[randomNumber])] = mOrganism
        }
        mUpdatePresenter.update(mField)
    }

    private fun findDirection(direction: Int): Boolean {
        val p = Field.getPointByDirection(mCurrent.x, mCurrent.y, direction)

        return mField.isInField(p.x, p.y) && mField[p.x, p.y] == null

    }
}
